core/src/test/java/org/elasticsearch/cluster/ClusterChangedEventTests.java

```diff
@@ -0,0 +1,350 @@
+     * based on what was set on the cluster state.
+     */
+    public void testLocalNodeIsMaster() {
+        final int numNodesInCluster = 3;
+        ClusterState previousState = createSimpleClusterState();
+        ClusterState newState = createState(numNodesInCluster, true, initialIndices);
+        ClusterChangedEvent event = new ClusterChangedEvent("_na_", newState, previousState);
+        assertTrue("local node should be master", event.localNodeMaster());
+
+        newState = createState(numNodesInCluster, false, initialIndices);
+        event = new ClusterChangedEvent("_na_", newState, previousState);
+        assertFalse("local node should not be master", event.localNodeMaster());
+    }
+
+    /**
+     * Test that the indices created and indices deleted lists between two cluster states
+     * are correct when there is no change in the cluster UUID.  Also tests metadata equality
+     * between cluster states.
+     */
+    public void testMetaDataChangesOnNoMasterChange() {
+        metaDataChangesCheck(false);
+    }
+
+    /**
+     * Test that the indices created and indices deleted lists between two cluster states
+     * are correct when there is a change in the cluster UUID.  Also tests metadata equality
+     * between cluster states.
+     */
+    public void testMetaDataChangesOnNewClusterUUID() {
+        metaDataChangesCheck(true);
+    }
+
+    /**
+     * Test the index metadata change check.
+     */
+    public void testIndexMetaDataChange() {
+        final int numNodesInCluster = 3;
+        final ClusterState originalState = createState(numNodesInCluster, randomBoolean(), initialIndices);
+        final ClusterState newState = originalState; // doesn't matter for this test, just need a non-null value
+        final ClusterChangedEvent event = new ClusterChangedEvent("_na_", originalState, newState);
+
+        // test when its not the same IndexMetaData
+        final String indexId = initialIndices.get(0);
+        final IndexMetaData originalIndexMeta = originalState.metaData().index(indexId);
+        // make sure the metadata is actually on the cluster state
+        assertNotNull("IndexMetaData for " + indexId + " should exist on the cluster state", originalIndexMeta);
+        IndexMetaData newIndexMeta = createIndexMetadata(indexId, originalIndexMeta.getVersion() + 1);
+        assertTrue("IndexMetaData with different version numbers must be considered changed", event.indexMetaDataChanged(newIndexMeta));
+
+        // test when it doesn't exist
+        newIndexMeta = createIndexMetadata("doesntexist");
+        assertTrue("IndexMetaData that didn't previously exist should be considered changed", event.indexMetaDataChanged(newIndexMeta));
+
+        // test when its the same IndexMetaData
+        assertFalse("IndexMetaData should be the same", event.indexMetaDataChanged(originalIndexMeta));
+    }
+
+    /**
+     * Test nodes removed check.
+     */
+    public void testNodesRemoved() {
```

Comment:
@bleskes Sorry you are correct, there is a `nodesAdded`.  What happened was, I originally took it out because I saw it was not used anywhere in core, until @dakrone pointed out that they could be used by plugins and shouldn't be removed.  So I will add a test in here for `nodesAdded` as well (all part of the same test) and include a scenario where nodes are both added and removed in the same cluster state update.